### PR TITLE
GS: Improve target size calcs and remove conservative framebuffer

### PR DESCRIPTION
**Pull request by stenzek**
### Description of Changes

Instead of using a fixed size for all render targets in the hardware renderer, this PR changes it to set the target size based on the FBW that the game configures. Unfortunately, we can't do the same for height, because it's not specified anywhere in the GS (only inferred based on the scissor height).

Using the scissor height alone would be a bit wasteful, because some games configure a stupidly high scissor, but then never draw anything in it. So, we try to be a bit smarter, and use the draw size instead. Naturally, this results in a variable size which changes throughout the frame, so we keep track of it with a FBP+FBW+PSM tuple, which is aged and eventually removed.

This leaves one nasty case which seems to be pretty common (e.g. NFS Carbon, GOW2, Burnout 3). They either do a recursive draw, or a shuffle with the depth buffer, with a scissor and draw height of 896. Because the framebuffer/depth buffer it's sampling from is usually only 448 pixels high, resizing the target to 896 would be kinda silly, because that area will just be rendered black, and waste bandwidth.

So, we try to detect these draws by looking at the gradient in the Y direction, and clamp the target size to the texture size in such cases. This works for everything I've checked so far, and results in a framebuffer size of 512x448 instead of 512x896 as was common before.

As an added bonus, because we're somewhat accurately predicting the target size, we can get rid of the conservative framebuffer option, since targets will only use as much VRAM as they need now. We can probably drop sparse texture too: it was only implemented in OpenGL anyway.

### Rationale behind Changes

Reducing VRAM usage when upscaling, performance increase.

Burnout 3 at 5x on a RX580 goes from 44fps to 85fps or so.

### Suggested Testing Steps

Check GS sensitive games.

### CODE REVIEW

**Comment by refractionpcsx2** [comment]
Nearly a clean sweep :) I went through all my dumps and they were all fine, except one.

Test Drive does some sort of huge depth buffer thing, I remember iMine having to fix it when he did the scaling stuff.

Heres a dump:
[testdrivehalfscreen.zip](https://github.com/PCSX2/pcsx2/files/9131209/testdrivehalfscreen.zip)

File: pcsx2/GameDatabase.cpp [hunk 1]
```diff
@@ -542,14 +537,6 @@ u32 GameDatabaseSchema::GameEntry::applyGSHardwareFixes(Pcsx2Config::GSOptions&
 				config.UserHacks_AutoFlush = (value > 0);
 				break;
 
-			case GSHWFixId::ConservativeFramebuffer:
-				config.ConservativeFramebuffer = (value > 0);
-				break;
-
-			case GSHWFixId::CPUFramebufferConversion:
-				config.UserHacks_CPUFBConversion = (value > 0);
-				break;
```
Comment (refractionpcsx2) [inline pcsx2/GameDatabase.cpp:551]: I presume you didn't mean to remove this? :)

**Comment by refractionpcsx2** [comment]
I know you hate WX, but might be worth cutting it out of there too :) I don't mind not adding new stuff to WX, but if we remove stuff we should really remove the option.

**Comment by coornio** [comment]
madlad at it once again <3

**Review by JordanTheToaster — APPROVED** [review]

**Review by lightningterror — APPROVED** [review]

**Comment by AhayriSG** [comment]
legendsflickering
https://github.com/PCSX2/pcsx2/issues/5244#issuecomment-1183800737

**Comment by stenzek** [comment]
That GS dump plays back fine for me on this PR, @refractionpcsx2 also checked.

**Comment by refractionpcsx2** [comment]
@AhayriSG did you even try it on this PR before posting that? cos two of us have tested it now and it works fine.

**Comment by AhayriSG** [comment]
> @AhayriSG did you even try it on this PR before posting that? cos two of us have tested it now and it works fine.
yep, build: 9377859
settings;
legendsflickering

**Review by JordanTheToaster — APPROVED** [review]

**Comment by Tyestor** [comment]
Keep up the awesome work sten

**Comment by refractionpcsx2** [comment]
[spider2.zip](https://github.com/PCSX2/pcsx2/files/9203169/spider2.zip)

Spider-man 2 has a bar coming out of his shadow (even in native) which isn't there in master

**Pull request merged**